src/core/App.ts

```diff
@@ -0,0 +1,285 @@
+        this.#onUserAction({ type: Action.AUTOTILE, layout: action.layout });
+        return;
+      case OverlayEventType.Visibility:
+        this.#hotkeyManager.setListeningGroups(action.visible
+          ? this.#globalKeyBindingGroups | DefaultKeyBindingGroups
+          : this.#globalKeyBindingGroups);
+        return;
+    }
+
+    // exhaustive switch-case guard
+    return ((): never => { })();
+  }
+
+  #onUserAction(action: HotkeyAction) {
+    // trivial delegation events
+    switch (action.type) {
+      case Action.TOGGLE:
+        this.#overlayManager.toggleOverlays();
+        return;
+      case Action.CANCEL:
+        this.#overlayManager.toggleOverlays(true);
+        return;
+      case Action.LOOP_GRID_SIZE:
+        this.#overlayManager.iteratePreset();
+        return;
+    }
+
+    const om = this.#overlayManager;
+    const dm = this.#desktopManager;
+    const window = dm.focusedWindow; if (!window) return;
+    const monitorIdx = om.activeMonitor ?? window.get_monitor();
+    const selection = om.getSelection(monitorIdx);
+
+    // events that require a window target
+    switch (action.type) {
+      case Action.CONFIRM:
+        if (selection) {
+          dm.applySelection(window, monitorIdx, om.gridSize, selection);
+          om.setSelection(null, monitorIdx);
+          this.#settings.get_boolean("auto-close") && om.toggleOverlays(true);
+        }
+        return;
+      case Action.PAN:
+        const curSel = selection ?? dm.windowToSelection(window, om.gridSize);
+        const newSel = pan(curSel, om.gridSize, action.dir);
+        om.setSelection(newSel, monitorIdx);
+        return;
+      case Action.ADJUST: {
+        const curSel = selection ?? dm.windowToSelection(window, om.gridSize);
+        const newSel = adjust(curSel, om.gridSize, action.dir, action.mode);
+        om.setSelection(newSel, monitorIdx);
+        return;
+      }
+      case Action.MOVE: {
+        dm.moveWindow(window, om.gridSize, action.dir);
+        return;
+      }
+      case Action.RESIZE: {
+        dm.resizeWindow(window, om.gridSize, action.dir, action.mode);
+        return;
+      }
+      case Action.GROW:
+        dm.autogrow(window);
+        return;
+      case Action.LOOP_PRESET: {
+        const preset = this.#getResizePreset(action.preset);
+        if (preset) {
+          const { gridSize, selection } = preset;
+          dm.applySelection(window, monitorIdx, gridSize, selection);
+        }
+        return;
+      }
+      case Action.RELOCATE:
+        dm.moveToMonitor(window);
+        return;
+      case Action.AUTOTILE:
+        if (action.layout === "main" || action.layout === "main-inverted") {
+          dm.autotile(AutoTileLayouts[action.layout], monitorIdx);
+        } else if (action.cols) {
+          dm.autotile(AutoTileLayouts[action.layout][action.cols], monitorIdx);
+        }
+        return;
+    }
+
+    // exhaustive switch-case guard
+    return ((): never => { })();
+  }
```

Comment:
```suggestion
      case Action.AUTOTILE:
        if (action.layout === "main" || action.layout === "main-inverted") {
          dm.autotile(AutoTileLayouts[action.layout], monitorIdx);
        } else if (action.cols) {
          dm.autotile(AutoTileLayouts[action.layout][action.cols], monitorIdx);
        }
        return;
      default:
        return ((): never => { })();
    }
  }
```

Wouldn't having a default case more this more idiomatic?

Reply:
I would argue its a matter of taste.